File: runtime/src/actions/var.jl
include("../types.jl")

# ActionVar runs the var action for runtime. Returns true if successful for testing
# purposes. Throws exception if error
function ActionVar(action::Action, vars::Dict)::Bool
    var_name = filter(d -> d.name == "var_name", action.args)[1]
    var_value = filter(d -> d.name == "var_value", action.args)[1]
    vars[var_name.value] = var_value.value
    return true
end

File: runtime/actions/var.go
package actions

import (
	"encoding/json"
	"log"

	"github.com/pitalco/clickscript/runtime/types"
)

type VarArgs struct {
	Name      string      `json:"name"`
	Value     interface{} `json:"value"`
	ValueType string      `json:"type"`
}

func Var(ctx types.Context, args []byte) error {
	var varArgs VarArgs
	// unmarshal the args for this action from bytes
	err := json.Unmarshal(args, &varArgs)
	if err != nil {
		log.Fatal(err)
	}
	// store the variable in memory as the specified type
	ctx.Memory[varArgs.Name] = varArgs.Value
	return nil
}
